tests/integration_tests/framework/wagon_build.py

```diff
@@ -0,0 +1,150 @@
+        return self.get_docker_client()
+
+    @property
+    def wagon_build_container(self):
+        return self._wagon_build_container
+
+    @property
+    def local_docker_images(self):
+        return [image.tags for image in self.docker_client.images.list()]
+
+    def _build_docker_image(self, dockerfile_object, image_name):
+        return self.docker_client.images.build(
+            fileobj=dockerfile_object,
+            tag=image_name)
+
+    def _run_docker_container(self, image_name, **kwargs):
+        """Start building the wagon."""
+        return self.docker_client.containers.run(image_name, **kwargs)
+
+
+class WagonBuilderMixin(DockerInterface):
+
+    @property
+    def wagon_build_time_limit(self):
+        """Max time it should take to build the wagon."""
+        return WAGON_BUILD_TIMEOUT
+
+    @property
+    def docker_image_name(self):
+        """The name of the Docker image that is used to build
+        the plugin wagon.
+        """
+        return WAGON_BUILD_DOCKER_IMAGE_NAME
+
+    @property
+    def plugin_root_directory(self):
+        """ Path to the plugin root directory."""
+        return None
+
+    @property
+    def docker_target_folder(self):
+        """Where to mount the docker local folder on the docker container.
+        """
+        return DOCKER_CONTAINER_BUILD_DIR
+
+    @property
+    def docker_volume_mapping(self):
+        """A volume mapping for mounting a local directory in a container.
+        The wagon will be here at the end of the build.
+        """
+        return {
+            self.plugin_root_directory: {
+                'bind': self.docker_target_folder,
+                'mode': 'rw'
+            }
+        }
+
+    def wagon_builder_image_ready(self):
+        """Check if the docker image already exists."""
+        return self.docker_image_name in self.local_docker_images
+
+    def build_docker_image(self):
+        """Build the wagon builder docker image."""
+        return self._build_docker_image(
+            dockerfile,
+            self.docker_image_name
+        )
+
+    def start_wagon_build(self):
+        """Start building the wagon."""
+        self._wagon_build_container = \
+            self._run_docker_container(
+                self.docker_image_name,
+                volumes=self.docker_volume_mapping,
+                detach=True)
+
+    def wagon_build_complete(self):
+        """Check if the wagon build is completed."""
+        if self.wagon_build_container.status == DOCKER_EXITED_STATUS:
+            return True
+        try:
+            log_message = \
+                self.wagon_build_container.logs().split('\n')[-2]
+        except (AttributeError, IndexError):
+            raise RuntimeError(
+                'Wagon builder container does not have logs'
+                'or they are not formatted as expected.')
+        if WAGON_FINISHED in log_message:
+            return True
+        return False
+
+    def wait_for_wagon_build(self):
```

Comment:
Seems to me that `while` is most appropriate since we are not performing an iteration. We are waiting for one of two conditions to be met. But I do think leaving if, elif un-terminated with a catch all condition will cause problems, so I will refactor.